### PR TITLE
table_helper: complete coroutinization

### PR DESCRIPTION
**Pull request by avikivity**
table_helper has some quite awkward code, improve it a little.

Code cleanup, so no reason to backport.

### CODE REVIEW

**Review by wmitros — CHANGES_REQUESTED** [review]
Looks mostly good but I found a small error

File: table_helper.hh [hunk 1]
```diff
@@ -41,7 +41,9 @@ private:
      * Should be changed alongside every _insert_stmt reassignment
      * */
     bool _is_fallback_stmt = false;
-
+private:
+    // Returns true is prepare succeeded, false if failed and there's still a chance to recover, exception on another type of error
```
[the remark this answers is not in the document]
Comment (avikivity) [reply table_helper.hh:45]: I'll update it.

File: table_helper.cc [hunk 2]
```diff
@@ -68,6 +68,31 @@ future<> table_helper::setup_table(cql3::query_processor& qp, service::migration
     } catch (...) {}
 }
 
+future<bool> table_helper::try_prepare(bool fallback, cql3::query_processor& qp, service::query_state& qs) {
+    auto& stmt = fallback ? _insert_cql_fallback : _insert_cql;
+    try {
+        shared_ptr<cql_transport::messages::result_message::prepared> msg_ptr = co_await qp.prepare(_insert_cql, qs.get_client_state());
```
[the remark this answers is not in the document]
Comment (avikivity) [reply table_helper.cc:74]: wow that's bad.

[the remark this answers is not in the document]
Comment (avikivity) [reply table_helper.cc:74]: It's practically impossible to trigger - you need to upgrade from such an old version that the table doesn't have the column and needs to fall back.

**Comment by avikivity** [comment]
v2:
 - improve comment on try_prepare()
 - changed try_prepare local `stmt` to be a string rather than an optional
 - use this `stmt` instead of trying to prepare _insert_cql in both paths

**Review by wmitros — APPROVED** [review]

**Comment by avikivity** [comment]
@denesb please review

File: table_helper.cc [hunk 3]
```diff
@@ -69,9 +69,10 @@ future<> table_helper::setup_table(cql3::query_processor& qp, service::migration
 }
 
 future<bool> table_helper::try_prepare(bool fallback, cql3::query_processor& qp, service::query_state& qs) {
-    auto& stmt = fallback ? _insert_cql_fallback : _insert_cql;
+    // Note: `_insert_cql_fallback` is known to be engaged if `fallback` is true, see cache_table_info below.
+    auto& stmt = fallback ? _insert_cql_fallback.value() : _insert_cql;
     try {
-        shared_ptr<cql_transport::messages::result_message::prepared> msg_ptr = co_await qp.prepare(_insert_cql, qs.get_client_state());
+        shared_ptr<cql_transport::messages::result_message::prepared> msg_ptr = co_await qp.prepare(stmt, qs.get_client_state());
```
Comment (denesb) [inline table_helper.cc:75]: The fix was merged into the wrong commit, this hunk belongs to the previous one.

Comment (avikivity) [reply table_helper.cc:75]: ouch, fixed

**Comment by avikivity** [comment]
v3:
  - moved previous review's fixed into the correct patch

**Comment by avikivity** [comment]
> [Test Result](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11007/testReport/) (3 failures / -37)
[Tests / Unit Tests / test_gossip_boot.test_gossip_boot.dev.1](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11007/testReport/junit/(root)/test_gossip_boot/Tests___Unit_Tests___test_gossip_boot_dev_1/)
[Tests / Unit Tests / test_gossip_boot.test_gossip_boot.dev.2](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11007/testReport/junit/(root)/test_gossip_boot/Tests___Unit_Tests___test_gossip_boot_dev_2/)
[Tests / Unit Tests / test_gossip_boot.test_gossip_boot.dev.3](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11007/testReport/junit/(root)/test_gossip_boot/Tests___Unit_Tests___test_gossip_boot_dev_3/)

Seen often, is this known?

**Comment by avikivity** [comment]
v4: rebased past CI instability

**Pull request merged**